### ISSUE TITLE
Reader view is too slow to be usable 

### ISSUE THREAD

**Issue by jdlongman** [burtonator/polar-bookshelf#844]
Reader is sluggish, has a terrible framerate. Scrolling through the page also feels disjointed. 
Also, scrolling through pages fast causes huge amounts of pop up or delay before content loads in. 
Overall it is unpleasant to use. Lag gives me motion sickness so this makes the whole app useless for me, since I have to read documents using a native PDF reader that doesn't lag anyway.
My computer runs Athril, Zathura, or even Adobe's reader, which while heavy, still isn't as bad as this. 
My computer has adequate specs, and every other reader works fine for me.
Specs:
Ubuntu 18
Intel i7
16GB RAM

**This should be a high priority issue. No amount of flashcard support or other gimmicks can make up for a poor experience of the core feature - which is READING!!**

**Comment by burtonator** [comment]
Other users have reported the same thing... Could you take a video/screen capture?
 
I think there might be TWO problem here.. Do you see a blank page with a little spinning progress wheel for about half a second and then the page renders?  

OR are you talking about just fast scrolling an existing pre-renedered page ?  

"This should be a high priority issue. No amount of flashcard support or other gimmicks can make up for a poor experience of the core feature - which is READING!!"

Yeah.. I agree..  But I think this might actually not be hitting everyone.  

Let me get more clarity here.  I might also try to get you to disable your GPU rendering in Polar (it's a flag that I need to figure out) and see if thats faster.  

Some users have issues with Electron on certain video cards.

Does the webapp exhibit the same behavior on the same PDF?

**Comment by jdlongman** [comment]
Indeed, there seems like two separate issues - the spinning progress wheel (or just white blank page) is the phenomenon produced from scrolling too quickly to pages further down. This is somewhat less urgent because after they load they stay there.

The more detrimental issue is the general high latency of scrolling through an (already loaded) page
It is difficult to adequately convey this experience through just a video, because it is more to do with the interactivity, you have to be scrolling the mouse and see the latency in practice to fully realize what is going on. The best way I can describe it is like using a web browser on an old computer without much RAM, it feels very similar to that. The movement of the page is inconsistent/jittery, and is nausea inducing.
My video card is Geforce GTX 990M, which while old, I still wouldn't consider obsolete.

If you want to reproduce this error I would suggest perhaps making a linux virtual machine with limited specs, you are sure to get the same experience. I have also tried it out on an old Sony Vaio with i3 processor and Windows 7, the same thing occurs.


I just tried out the webapp and it seems to be a little more smooth than the desktop app, however I used it side by side with my native pdf readers, and it definitely feels like it is a latency issue. That is, when I scroll in an app like Athril, it scrolls instantly, but there feels like there is a delay before the scroll kicks in in Polar. So maybe there is a parameter that can be tuned to fix this, since its not like Polar is taking up 100% of my CPU or RAM.

**Comment by jdlongman** [comment]
Revision to the above: scrolling up-down-up-down-up-down very fast spikes the CPU usage to 90% for Polar. I tried the same thing in Atril and it got 8% CPU usage, Zathura was 16% CPU usage. In all cases memory is around 1.5~2% usage.

**Comment by burtonator** [comment]
Thanks what does the cpu look like for the browser version.

I think this might be related to an electron gpu bug which i don't think impacts chrome.

**Comment by jdlongman** [comment]
For Firefox I'm getting 30-40%
For Chrome its 50-70%

**Comment by burtonator** [comment]
ok.. so FF and chrome aren't nailing the CPU...

could you try running

export POLAR_DISABLE_HARDWARE_ACCELERATION=true

from the command line and starting polar manually?

it the argument worked it will print out "Disabling hardware acceleration"

... this just disables it for the Electron process .. not globally or anything. 

If it's faster then that's the problem.

I just don't know a way to detect it and enable it.

another option is to try pdf.js  "raw" without any polar extensions here:

https://mozilla.github.io/pdf.js/web/viewer.html

and see if it's much faster.

I suspect its' the GPU though :-/

**Comment by jdlongman** [comment]
Doing the above export spikes CPU to 121%

**Comment by jdlongman** [comment]
I also have an embedded intel GPU, is there an export I can use to revert to using that?

**Comment by burtonator** [comment]
no... you can only disable... using a GPU altogether... I think ther'es not much we can do and  you're just getting hit with the GPU bug..

The fact that FF + Chrome are < 100% means that the GPU is handling it properly.

IS this URL still slow for you?

https://mozilla.github.io/pdf.js/web/viewer.html

**Comment by burtonator** [comment]
On my machine.. it's insanely fast.  My box is about 2-3 years old with a moderate GPU.

Do you have a recent version of ubuntu running?

**Comment by jdlongman** [comment]
The pdf.js link has the same laggy feel as the webviewer.
I am not sure you will be able to see what I mean unless you do it side-by-side with a native pdf reader like Atril. From the moment I scroll to the actual page movement there is definitely a finite pause compared to native readers.

**Comment by jdlongman** [comment]
I am on Ubuntu 18.1.

**Comment by jdlongman** [comment]
Also take into consideration that the same thing may indeed be happening on your machine but you are not perceiving it. It is similar as those who get motion sickness from VR lag and those who don't perceive it.
Perhaps there is some kind of testing tool that can measure the latency between scroll and move so that you can objectively measure the difference.

**Comment by jdlongman** [comment]
Additionally, you should try it out on a virtual machine, that way you can change the specs and test for multiple device configurations, not just your own computer, which probably has above average specs.

**Comment by burtonator** [comment]
@jdlongman 18.1 or 18.10 ? 

Good point about the perception issue. This would be a tough bug to debug... 

It seems like pdf.js just in general doesn't work for you though. We're definitely getting people complaining about this just not sure what % it impacts.

What if I reworked it so that it scrolled in large increments? 

One idea could be to try on another machine you have access to and see if you see the same issues. Maybe even a MacOS or Windows machine to see if you get lagginess.  Would really appreciate your feedback here because if it's just a hardware issue or OS issue I can't really resolve that too much but if it's a software issue I have more control over a fix.

**Comment by burtonator** [comment]
I might also investigate whether web workers (or threads) are working properly. If for some reason it's not using them that might explain why it's slow rendering as the page loading would be done in the UI thread and lock up for you but on more modern or faster machines you might not see the issue.

**Comment by burtonator** [comment]
well the worker is definitely loaded...

**Comment by jdlongman** [comment]
I did mention above that on an older Windows 7 device with i3 processor the same thing happens. In fact, it is even worse.
Increasing the scroll increment would not fix the issue, it would exacerbate it because of the jerkiness. What needs to be fixed is the **latency** between mouse scroll and page scroll.

You can debug it by measuring the latency between mouse movement and page movement, under different specs of virtual machine.

**Comment by burtonator** [comment]
ok... I think  I understand it a bit better... There's a new version of Electron out (5x) and we're going to jump from 3.x to 5.x and that might help fix the problem possibly.  

but if you're seeing it with pdf.js it makes me think its core to pdf.js or the browser which isn't a good sign.

Does FF + that pdf.js URL exhibit the problem for you?

I tried on okular and they both seemed about the same but maybe my machine is too fast and I should try a VM...

**Comment by burtonator** [comment]
Do you see it with just the scroll wheel or dragging the sidebar too?

I think I see a bit of it with the scroll wheel but not much when dragging the sidebar.

**Comment by jdlongman** [comment]
I was trying it with firefox already so yes the same problem.

Dragging the sidebar seems less susceptible to the problem, since there is no inertia.

**Comment by burtonator** [comment]
interesting... wonder if it's the scrolling algorithm itself being wonky and I wonder if I can disable that.

Another idea.  Could you take a video on a modern phone and upload it?

Even if the phone is only 30FPS I could be able to see it (IF I detect it).

300-200ms is about the time a human can detect latency and 1/30th of a second is 30ms ... 

We're going to jump to Electron 5.x here soon and that might also help but not sure of course.

**Comment by burtonator** [comment]
https://bugs.launchpad.net/ubuntu/+source/nvidia-graphics-drivers-390/+bug/1755182

does that seem like your issue?

**Comment by burtonator** [comment]
https://electronjs.org/docs/api/sandbox-option

this might be another thing you could try.. you basically need to take polar and add the --enable-sandbox option to the 'electron' process.

Might be best to build from source if  you can try that... I might add this option as a STANDARD option considering its faster since stability is less of an issue since we control all the content and aren't running video or audio or webgl or anything that could be unstable.

**Comment by burtonator** [comment]
Another related issue:

https://bugs.launchpad.net/ubuntu/+source/chromium-browser/+bug/1768199/

**Comment by burtonator** [comment]
Seems like Electron 5.x will fix this problem.. I can probably upgrade to 4.x soon but 5.x is probably 1-2 weeks away.  

If this is really hurting you I could create a branch for you to run. I think we have one already but it's not running master so might be unstable.  Seems to work just fine with Electron 5.x

**Comment by jdlongman** [comment]
The second of the above links may be related. The first one talks about non-specific freezing, and does not seem like the same problem. 
I will test the Electron 5 version and let you know.

**Comment by jdlongman** [comment]
Regarding taking video: the problem is not the FPS, the problem is that you won't observe it in the video because the problem is **interaction** based. You have to actually move the mouse yourself to feel it.

**Comment by jdlongman** [comment]
![compare](https://user-images.githubusercontent.com/48571943/56497644-67570780-64f6-11e9-95dd-6008f27f7701.gif)
Here I have attempted to show this phenomenon. First I scroll fast on a document in Atril, then do the same movements in Polar.

**Comment by burtonator** [comment]
Need a chrome later than: 

Newer Chromium (66.0.3359.139

**Comment by burtonator** [comment]
I'm going to push 1.18 with Electron 4.x so would appreciate letting me know if it fixes your issue.

**Comment by jdlongman** [comment]
Ok, looking forward to it!

**Comment by jdlongman** [comment]
Tested 1.17.5 on Windows 8.1 on a more recent computer.
It seems that problem 1 (general scroll lag within page) is reduced compared to Linux. However problem 2 (jittering as scrolling across page boundaries) is still quite prevalent.

**Comment by burtonator** [comment]
Thanks.. I have a fix which will be in 1.17.6 which definitely improves performance on large PDFs and that also might better highlight the glitch issue and created another bug for it as I"m not sure if these are related.

Still hoping that Electron 5.x helps fix this but it's not out of beta yet.

**Comment by jdlongman** [comment]
problem persists as of 1.18.1.

**Comment by burtonator** [comment]
Gasp... well at least we can rule that out! I'm also seeing some latency on my end for larger files too and I"ll try to track it down a bit more.
